### PR TITLE
vdk-jobs-troubleshooting: Add thread-dump utility

### PR DESCRIPTION
**Pull request by doks5**
This change adds the implementation of a thread-dump utility to the vdk-jobs-troubleshooting plugin.

The utility uses an http server, through which an administrator is able to force a stacktrace dump of all threads used by the python process of the data job. The server is bound to a port on the localhost, so to get the stacktrace, one needs to be attached to the data job pod.

Testing Done: Added unit tests for the utility registry, and tested the plugin itself locally by running a simple data job and examining the execution logs.

Signed-off-by: Andon Andonov <andonova@vmware.com>

### CODE REVIEW

File: projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/healthcheck_server.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
+# Copyright 2021 VMware, Inc.
+# SPDX-License-Identifier: Apache-2.0
+from http.server import HTTPServer
+from http.server import SimpleHTTPRequestHandler
+from threading import Thread
+from typing import Any
+
+
+class HealthCheckServer:
```
Comment (tozka) [inline projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/healthcheck_server.py:9]: Provide descripton of the class purpose, usage, examples

File: projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/thread_dump.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    TROUBLESHOOT_PORT_TO_USE,
+)
+from vdk.plugin.jobs_troubleshoot.troubleshoot_utilities.healthcheck_server import (
+    HealthCheckServer,
+)
+
+log = logging.getLogger(__name__)
+
+
+class ThreadDumpHandler(BaseHTTPRequestHandler):
```
Comment (tozka) [inline projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/thread_dump.py:22]: Provide descripton of the class purpose, usage, examples

File: projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/jobs_troubleshoot_plugin.py [hunk 3]
```diff
@@ -5,21 +5,59 @@
...
+from vdk.plugin.jobs_troubleshoot.troubleshoot_utilities.utilities_registry import (
+    get_utilities_to_use,
+)
 
 log = logging.getLogger(__name__)
 
 
 class JobTroubleshootingPlugin:
+    def __init__(self):
+        self.troubleshooting_utils: Optional[List[ITroubleshootUtility]] = []
```
Comment (tozka) [inline projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/jobs_troubleshoot_plugin.py:25]: Please take into account https://github.com/vmware/versatile-data-kit/wiki/Coding-Standard#clarification-on-access-modifiers-in-python

**Review by tozka — COMMENTED** [review]
Please write functional test . The cookiecutter provides template - https://github.com/tozka/cookiecutter-vdk-plugin/blob/main/vdk-%7B%7Bcookiecutter.plugin_name%7D%7D/tests/test_plugin.py 

The functional tests are very powerful mechanism to make sure the functionality we write really works as expected - as they are really "data jobs" that use that functionality.

File: projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/jobs_troubleshoot_plugin.py (hunk 3 above)
Comment (murphp15) [inline projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/jobs_troubleshoot_plugin.py:25]: Do you need optional? 
Can empty list just represent none?

File: projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/thread_dump.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            self._log_thread_dump()
+        else:
+            self.send_error(404)
+
+    @staticmethod
+    def _log_thread_dump():
+        try:
+            for t in threading.enumerate():
+                log.info(
+                    f"Thread:{t.getName()} alive:{t.is_alive()} daemon:{t.isDaemon()}"
```
Comment (murphp15) [inline projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/thread_dump.py:37]: Does the name here match the threadId in the loop below?

File: projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/utilities_registry.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    """
+    The troubleshooting utilities registry is where all utility objects are to
+    be initialized.
+    TODO: Come up with a more elegant approach to register utilities.
+
+    :param job_config: The data job configuration.
+    :return: A dictionary with all available troubleshooting utilities.
+    """
+    registered_utilities: Dict[str, Any] = {}
+    registered_utilities["thread-dump"] = ThreadDumpUtility(
```
Comment (murphp15) [inline projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/utilities_registry.py:32]: So it is always enabled?

**Comment by dakodakov** [comment]
> Please write functional test . The cookiecutter provides template - https://github.com/tozka/cookiecutter-vdk-plugin/blob/main/vdk-%7B%7Bcookiecutter.plugin_name%7D%7D/tests/test_plugin.py
> 
> The functional tests are very powerful mechanism to make sure the functionality we write really works as expected - as they are really "data jobs" that use that functionality.

I just added a functional test.

File: projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/thread_dump.py (hunk 4 above)
Comment (dakodakov) [reply projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/thread_dump.py:40]: yes..

File: projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/thread_dump.py (hunk 2 above)
Comment (dakodakov) [reply projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/thread_dump.py:21]: Done.

File: projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/healthcheck_server.py (hunk 1 above)
Comment (dakodakov) [reply projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/healthcheck_server.py:9]: Done.

File: projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/jobs_troubleshoot_plugin.py (hunk 3 above)
Comment (dakodakov) [reply projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/jobs_troubleshoot_plugin.py]: Fixed.

File: projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/healthcheck_server.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+        Parameters:
+            port (int): The port number on which the server will listen for requests.
+            handler (Any, optional): The request handler class. Defaults to SimpleHTTPRequestHandler.
+        """
+
+        if handler:
+            self._server = HTTPServer(("", port), handler)
+        else:
+            self._server = HTTPServer(("", port), SimpleHTTPRequestHandler)
```
Comment (murphp15) [inline projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/healthcheck_server.py:26]: Why would you start it up without the thread dump handler? what would this do?

File: projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/healthcheck_server.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        self._thread = Thread(target=self._server.serve_forever)
+
+    def __enter__(self):
```
Comment (murphp15) [inline projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/healthcheck_server.py:29]: I don't see that this is used?
Can we just remove it?

**Comment by murphp15** [comment]
I think that there should be docs describing how to use it. 
Especially to emphasise that the onus is on the developer to open a proxy to the pod if they want to use this functionality.

**Review by murphp15 — APPROVED** [review]

**Review by tozka — APPROVED** [review]
Looks good to me.

File: projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/healthcheck_server.py (hunk 6 above)
Comment (dakodakov) [reply projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/healthcheck_server.py]: agreed. added an error log instead.

File: projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/healthcheck_server.py (hunk 7 above)
Comment (dakodakov) [reply projects/vdk-plugins/vdk-jobs-troubleshooting/src/vdk/plugin/jobs_troubleshoot/troubleshoot_utilities/healthcheck_server.py]: Agreed. Removed.

**Comment by tozka** [comment]
I can see it's not release yet ([the release job did not trigger](https://gitlab.com/vmware-analytics/versatile-data-kit/-/pipelines/739346301) ) . It's commented out but I think you can release it: 
https://github.com/vmware/versatile-data-kit/blob/main/projects/vdk-plugins/vdk-jobs-troubleshooting/.plugin-ci.yml#L33

**Pull request merged**